docs/design_proposals/sync-improvements.md

```diff
@@ -0,0 +1,319 @@
+  - Transplant a folder structure, or flatten the folder structure
+- Offer a way to use destination inference.
+- Avoid ambiguity
+
+#### Problems with existing syntax
+- The flattening of the file structure at the destination is surprising. For example:
+  ```yaml
+  sync:
+    'src/**/*.js': dest/
+  ```
+  will put all `.js` files in the `dest` folder.
+- The triple-star syntax is is quite implicit and a surprising extension of the standard globbing syntax. For example:
+  ```yaml
+  sync:
+    'src/b/c/***/*.py': dest/
+  ```
+  will recreate subfolders _below_ `src/b/c` at `dest/` and sync `.py` files.
+  The triple-star therefore triggers two actions which should not depend on each other:
+  - do not flatten the directory structure
+  - strip everything on the source path up to `***`
+
+#### Suggestion
+These above problems can be made less surprising by a more explicit configuration syntax.
+The `artifact.sync` will be upgraded to a list of *sync rules*.
+
+```yaml
+sync:
+# Existing use-case: Copy all js files under 'src' flat into 'app'
+# This corresponds to the current `'src/**/*.js': app/`
+- from: 'src/**/*.js'
+  to: app/
+  flatten: true  # default to false
+  
+# Existing use-case: Copy all js under 'src' into 'dest', re-creating the directory structure below 'src'  
+# This corresponds to the current `'src/***/*.js': dest/`
+- from: 'src/**/*.js'
+  to: dest/
+  strip: 'src'   # default to ""
+  
+# New use-case: Copy all python files and infer their destination.
+- from: '**/*.py'
+  inferTo: true   # <- tbd, default to false
+```
+
+When determining the destination for a given file, all sync rules will be considered.
+Thus, a single file may be copied to multiple destinations.
+
+Error cases:
+- It should be an error to specify `inferTo=true` together with `to` in the same sync rule.
+  ```yaml
+  - from: '*.js'
+    to: dest/
+    inferTo: true
+  ```
+- It should be an error to specify either of `flatten` or `strip` together with `inferTo`.
+  ```yaml
+  - from: '*.js'
+    inferTo: true
+    flatten: true        # ERR
+  - from: 'src/**/*.js'
+    inferTo: true
+    strip: src/          # ERR
+  ```
+- It should be an error to specify `strip` with a prefix which which does not match the static prefix of the `from` field.
+  ```yaml
+  - from: 'src/**/*.js'
+    to: dest/
+    strip: src/sub        # ERR
+  - from: 'src/**/*.js'
+    to: dest/
+    strip: app/           # ERR
+  ```
+
+Open questions about the configuration:
+
+**\<Is the syntax clear enough?\>**
+Is the syntax clear?
+Is the configuration intuitive to use?
+Is there any ambiguity which needs attention?
+
+Resolution: __Not Yet Resolved__
+
+**\<Does it improve clarity to have the `inferTo` field?\>**
+A different approach could be to fall back to inference, if the `to` field is left blank.
+This would have the advantage that there needs to be no validation about having `to` and `inferTo` in the same rule, which should cause an error.
```

Comment:
I prefer this approach: it seems pretty intuitive that a missing `to` means _fill me in_.